tsdb/agent/db.go

```diff
@@ -646,56 +660,67 @@ func (db *DB) Close() error {
 type appender struct {
 	*DB
 
-	pendingSeries  []record.RefSeries
-	pendingSamples []record.RefSample
+	pendingSeries    []record.RefSeries
+	pendingSamples   []record.RefSample
+	pendingExamplars []record.RefExemplar
 }
 
 func (a *appender) Append(ref uint64, l labels.Labels, t int64, v float64) (uint64, error) {
-	if ref == 0 {
-		return a.Add(l, t, v)
-	}
-	return ref, a.AddFast(ref, t, v)
-}
+	series := a.series.GetByID(ref)
+	if series == nil {
+		// Ensure no empty or duplicate labels have gotten through. This mirrors the
+		// equivalent validation code in the TSDB's headAppender.
+		l = l.WithoutEmpty()
+		if len(l) == 0 {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
+		}
 
-func (a *appender) Add(l labels.Labels, t int64, v float64) (uint64, error) {
-	hash := l.Hash()
-	series := a.series.GetByHash(hash, l)
-	if series != nil {
-		return series.ref, a.AddFast(series.ref, t, v)
-	}
+		if lbl, dup := l.HasDuplicateLabelNames(); dup {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+		}
 
-	// Ensure no empty or duplicate labels have gotten through. This mirrors the
-	// equivalent validation code in the TSDB's headAppender.
-	l = l.WithoutEmpty()
-	if len(l) == 0 {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
-	}
+		var created bool
+		series, created = a.getOrCreate(l)
+		if created {
+			a.pendingSeries = append(a.pendingSeries, record.RefSeries{
+				Ref:    series.ref,
+				Labels: l,
+			})
 
-	if lbl, dup := l.HasDuplicateLabelNames(); dup {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+			a.metrics.numActiveSeries.Inc()
```

Comment:
The agent increases a total_created_series counter here, alongside decrementing a total_deleted_series counter on GC. Do we want to port over those two metrics?

Reply:
decrementing a total_deleted_series counter

counters can't be decreased, what do you mean?

Reply:
Ha, sorry, increasing a total_deleted_series counter.

grafana/agent code: https://github.com/grafana/agent/blob/718c330500f1ac0f0e6734bbf049ea3713f360fa/pkg/metrics/wal/wal.go#L442

We added this at some point to do `rate` with series going in/out of memory. 

Note that Prometheus Agent already has a gauge for series that are currently pending deletion:
https://github.com/prometheus/prometheus/blob/4bdaea7663264d165b7990e0c8573e9334803cd5/tsdb/agent/db.go#L574